L1Trigger/CSCTriggerPrimitives/src/CSCGEMMotherboard.h

```diff
@@ -0,0 +1,450 @@
+
+  template <class T>  
+  void correlateLCTsGEM(const T& best, const T& second, const GEMCoPadDigiIds& coPads, 
+			CSCCorrelatedLCTDigi& lct1, CSCCorrelatedLCTDigi& lct2, enum CSCPart);
+
+  // specific functions
+  CSCCorrelatedLCTDigi constructLCTsGEM(const CSCALCTDigi& alct, const GEMCoPadDigi& gem, enum CSCPart, int i);
+  CSCCorrelatedLCTDigi constructLCTsGEM(const CSCCLCTDigi& clct, const GEMCoPadDigi& gem, enum CSCPart, int i);
+  CSCCorrelatedLCTDigi constructLCTsGEM(const CSCALCTDigi& alct, const CSCCLCTDigi& clct,
+					const GEMCoPadDigi& gem, enum CSCPart p, int i);
+  CSCCorrelatedLCTDigi constructLCTsGEM(const CSCALCTDigi& alct, const CSCCLCTDigi& clct,
+					const GEMPadDigi& gem, enum CSCPart p, int i);
+  /*
+   * General function to construct integrated stubs from CSC and GEM information.
+   * Options are:
+   * 1. ALCT-CLCT-GEMPad
+   * 2. ALCT-CLCT-GEMCoPad
+   * 3. ALCT-GEMCoPad
+   * 4. CLCT-GEMCoPad
+   */
+  CSCCorrelatedLCTDigi constructLCTsGEM(const CSCALCTDigi& alct, const CSCCLCTDigi& clct,
+					const GEMPadDigi& gem1,  const GEMCoPadDigi& gem2,
+					enum CSCPart p, int i);
+  
+  void retrieveGEMPads(const GEMPadDigiCollection* pads, unsigned id);
+  void retrieveGEMCoPads();
+
+  template <class T>  
+  unsigned int findQualityGEM(const CSCALCTDigi&, const CSCCLCTDigi&);
+
+  template <class T>  
+  void printGEMTriggerPads(int bx_start, int bx_stop, enum CSCPart);
+
+  bool isPadInOverlap(int roll);
+  
+  void setupGeometry();
+
+  /** Chamber id (trigger-type labels). */
+  unsigned gemId; 
+
+  const GEMGeometry* gem_g;
+  bool gemGeometryAvailable;
+
+  std::vector<GEMCoPadDigi> gemCoPadV;
+
+  // map< bx , vector<gemid, pad> >
+  GEMPadDigiIdsBX pads_;
+  GEMCoPadDigiIdsBX coPads_;
+
+  //  deltas used to match to GEM pads
+  int maxDeltaBXPad_;
+  int maxDeltaBXCoPad_;
+  int maxDeltaPadL1_;
+  int maxDeltaPadL2_;
+
+  // send LCT old dataformat
+  bool useOldLCTDataFormat_;
+
+  // promote ALCT-GEM pattern
+  bool promoteALCTGEMpattern_;
+
+  bool promoteALCTGEMquality_;
+  bool promoteCLCTGEMquality_;
+
+  // LCT ghostbusting
+  bool doLCTGhostBustingWithGEMs_;
+};
+
+
+template <class S>  
+void CSCGEMMotherboard::correlateLCTsGEM(const S& bestLCT,
+					 const S& secondLCT,
+					 const GEMCoPadDigiIds& coPads, 
+					 CSCCorrelatedLCTDigi& lct1,
+					 CSCCorrelatedLCTDigi& lct2, 
+					 enum CSCPart p)
+{
+  bool bestValid     = bestLCT.isValid();
+  bool secondValid   = secondLCT.isValid();
+
+  // get best matching copad1
+  GEMCoPadDigi bestCoPad = bestMatchingPad<GEMCoPadDigi>(bestLCT, coPads, p);
+  GEMCoPadDigi secondCoPad = bestMatchingPad<GEMCoPadDigi>(secondLCT, coPads, p);
+
+  if (bestValid and !secondValid) secondLCT = bestLCT;
+  if (!bestValid and secondValid) bestLCT   = secondLCT;
+
+  bool lct_trig_enable;
+  if (std::is_same<S, CSCALCTDigi>::value) lct_trig_enable = alct_trig_enable;
+  if (std::is_same<S, CSCCLCTDigi>::value) lct_trig_enable = clct_trig_enable;
```

Comment:
it seems this would be better handled by template specialization? If nothing else, this construct breaks if someone uses a type that is not these two kinds (silently). If there are only two supported types, then that should be a compile time error